bs_detector/bias.py

```diff
@@ -475,27 +483,59 @@ def print_raw_data_for_features(list_of_sentences):
             print(feat.keys())
             KEYS_DONE = True
         print(feat.values())
-#print_raw_data_for_features(get_list_from_file('input_text_original'))
+
+
+# print_raw_data_for_features(get_list_from_file('input_text_original'))
 
 def compute_bias(sentence_text):
-    features = extract_bias_features(unicode(sentence_text, errors='ignore'))
+    if isinstance(sentence_text, unicode):
+        features = extract_bias_features(sentence_text)
+    else:
+        features = extract_bias_features(unicode(sentence_text, errors='ignore'))
     BS_SCORE = (-0.5581467 +
-          0.3477007 * features['vader_sentiment'] +
-          -2.0461103 * features['opinion_rto'] +
-          0.5164345 * features['modality'] +
-          8.3551389 * features['liwc_3pp_rto'] +
-          4.5965115 * features['liwc_tent_rto'] +
-          5.737545 * features['liwc_achiev_rto'] +
-          5.6573254 * features['liwc_discr_rto'] +
-          -0.953181 * features['bias_rto'] +
-          9.811681 * features['liwc_work_rto'] +
-          -16.6359498 * features['factive_rto'] +
-          3.059548 * features['hedge_rto'] +
-          -3.5770891 * features['assertive_rto'] +
-          5.0959142 * features['subj_strong_rto'] +
-          4.872367 * features['subj_weak_rto'])
+                0.3477007 * features['vader_sentiment'] +
+                -2.0461103 * features['opinion_rto'] +
+                0.5164345 * features['modality'] +
+                8.3551389 * features['liwc_3pp_rto'] +
+                4.5965115 * features['liwc_tent_rto'] +
+                5.737545 * features['liwc_achiev_rto'] +
+                5.6573254 * features['liwc_discr_rto'] +
+                -0.953181 * features['bias_rto'] +
+                9.811681 * features['liwc_work_rto'] +
+                -16.6359498 * features['factive_rto'] +
+                3.059548 * features['hedge_rto'] +
+                -3.5770891 * features['assertive_rto'] +
+                5.0959142 * features['subj_strong_rto'] +
+                4.872367 * features['subj_weak_rto'])
     return BS_SCORE
 
+
+@contextmanager
+def poolcontext(*args, **kwargs):
+    pool = multiprocessing.Pool(*args, **kwargs)
+    yield pool
+    pool.terminate()
+
+
+def compute_statement_bias(statement_text, n_jobs=1):
+    sentences = nltk.sent_tokenize(statement_text.decode("ascii", "ignore"))
+    max_len = max(map(len, sentences))
+
+    if len(sentences) == 0:
+        return 0
+
+    avg_bias = 0
+
+    with poolcontext(processes=n_jobs) as pool:
+        bs_scores = pool.map(compute_bias, sentences)
+        avg_bias = sum(bs_scores)
```

Comment:
Using multiprocessing here will only be faster if the number of sentences is `yoooge`. Have you done any profiling to see if it is necessary on reasonably sized documents?

I would rather make the code more efficient that throw processors at the problem if we can.

Reply:
No, no profiling yet but thanks for suggesting, agree. The rationale was that the user could decide and edit the n_jobs parameter to suit their needs. Better to make the option available. I think we can have this and also continue to make more efficient code so no need to choose one or the other. But so are you saying that you don't think an article with 200 sentences wouldn't see a performance increase over serial? I tend to think there would be some performance increase on a multi-core system. And what do you consider yooge? 1K 1M 10M?

Reply:
Yeah I think 1K is the minimum at which I would parallelize. You have to copy all the data into the subprocesses which should take a big chunk of the time in processing them. If you make a process pool with one job does it actually make a subprocess and do all the overhead? Or do they use a dummy process pool that does the work in the main process?

Reply:
Of course I would want to profile and find the hot spots and optimize them, before I started to parallelize within a single document. It is much more efficient to parallelize over the independent documents if that is possible.

Reply:
Ok so, I ran profiling and I do see a performance improvement in terms of total time. See attached. Thoughts? I've also created a new non-MP method in the process, so we'll have that either way now.

MP (7494ms)
<img width="1153" alt="compute_statement_mp" src="https://user-images.githubusercontent.com/25912784/33329519-be3278b8-d42a-11e7-9895-1dfa57b32ad1.png">

No MP (22179ms)
<img width="1296" alt="compute_statement_nomp" src="https://user-images.githubusercontent.com/25912784/33329524-c429b97a-d42a-11e7-92f1-53fde9d2c0db.png">

Reply:
Oh, also, I think the parallelization could benefit at the document and also the within document levels since sentences are independent units of bias scoring and if we agree on the above interpretation then there are gains to be made with MP on statement-level (a document can be thought of in some sense as a big statement)

Reply:
I am pleasantly surprised about that 3x on (i assume 4 cores). I see an easy pick up in accelerating the serial implementation.